File: scripts/dump-ranks.mjs
// G-04 二実装照合の TS 側。Python 側(harness/reference/dump_ranks.py)と同じ形で吐く。
//
// 照合の対象を三層に分ける。上の層が壊れていると下の層の一致は意味を持たない。
//   1. 分かち書き …… 全 13,600 チャンク + 全 2,808 クエリのバイグラムを網羅照合(sha256)
//   2. 疎索引     …… N / avgdl / 語数 / 文書長の総和と checksum / df の全語照合
//   3. 検索順位   …… 疎 200 問・密 20 問・融合 20 問の上位 20 件

import fs from "node:fs";
import crypto from "node:crypto";
import { resolve, dirname } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { bigrams, tokenLength } from "../src/lib/tokenize.mjs";
import { buildIndex, search as searchSparse } from "../src/lib/bm25.mjs";
import { searchDense } from "../src/lib/dense.mjs";
import { rrfFuse } from "../src/lib/metrics.mjs";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const DIMS = 384;
export const SPARSE_QUERIES = 200;
export const DENSE_QUERIES = 20;
export const TOPK = 20;

const sha = (s) => crypto.createHash("sha256").update(s, "utf8").digest("hex").slice(0, 16);

/** 等間隔で選ぶ。両実装で同じ問いを見るため、乱数を使わない */
export function stride(n, want) {
  const step = Math.max(1, Math.floor(n / want));
  const out = [];
  for (let i = 0; i < n && out.length < want; i += step) out.push(i);
  return out;
}

async function main() {
  const { chunks } = JSON.parse(fs.readFileSync(resolve(ROOT, "data/chunks.json"), "utf8"));
  const { items } = JSON.parse(fs.readFileSync(resolve(ROOT, "data/oracle.json"), "utf8"));
  const bin = fs.readFileSync(resolve(ROOT, "data/index/vec_f32.bin"));
  const vecs = new Float32Array(bin.buffer, bin.byteOffset, bin.byteLength / 4);
  const qbin = fs.readFileSync(resolve(ROOT, "data/index/query_f32.bin"));
  const qvecs = new Float32Array(qbin.buffer, qbin.byteOffset, qbin.byteLength / 4);
  const ids = chunks.map((c) => c.id);

  // --- 層 1: 分かち書き(網羅) ---
  const tokenHashes = {
    chunks: chunks.map((c) => sha([...bigrams(c.indexText)].join(""))),
    queries: items.map((it) => sha([...bigrams(it.query)].join(""))),
    lengths: sha(chunks.map((c) => tokenLength(c.indexText)).join(",")),
  };

  // --- 層 2: 疎索引 ---
  const ix = buildIndex(chunks.map((c) => ({ id: c.id, text: c.indexText })));
  const terms = [...ix.postings.keys()].sort();
  const sparseIndex = {
    N: ix.N,
    avgdl: ix.avgdl,
    termCount: terms.length,
    dlSum: [...ix.dl].reduce((a, b) => a + b, 0),
    termsHash: sha(terms.join("")),
    dfHash: sha(terms.map((t) => ix.postings.get(t).length / 2).join(",")),
  };

  // --- 層 3: 検索順位 ---
  const sIdx = stride(items.length, SPARSE_QUERIES);
  const dIdx = stride(items.length, DENSE_QUERIES);
  const sparseRanks = sIdx.map((i) => ({
    item: items[i].id,
    ids: searchSparse(ix, items[i].query, {
      topK: TOPK,
      exclude: new Set(items[i].excludeFromResults),
    }).map((h) => h.id),
  }));
  const denseRanks = [];
  const hybridRanks = [];
  for (const i of dIdx) {
    const it = items[i];
    const exclude = new Set(it.excludeFromResults);
    const q = qvecs.subarray(i * DIMS, (i + 1) * DIMS);
    const d = searchDense(vecs, DIMS, ids, q, { topK: TOPK, exclude });
    const s = searchSparse(ix, it.query, { topK: TOPK, exclude });
    denseRanks.push({ item: it.id, ids: d.map((h) => h.id), topScore: d[0]?.score ?? null });
    hybridRanks.push({ item: it.id, ids: rrfFuse([d, s], { topK: TOPK }).map((h) => h.id) });
  }

  const out = { impl: "ts", tokenHashes, sparseIndex, sparseRanks, denseRanks, hybridRanks };
  fs.writeFileSync(resolve(ROOT, "data/ranks-ts.json"), JSON.stringify(out));
  console.log(
    `TS 側を書き出した: チャンク ${tokenHashes.chunks.length} / クエリ ${tokenHashes.queries.length} / ` +
      `疎 ${sparseRanks.length} 問 / 密 ${denseRanks.length} 問`
  );
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) await main();

File: harness/reference/dump_ranks.py
"""G-04 二実装照合の Python 側。scripts/dump-ranks.mjs と同じ形で吐く。

TS 側の出力は読まない。**同じ入力から独立に計算する**。
読んでしまうと照合ではなく写経になる。
"""

from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
import tenkyo as T  # noqa: E402

ROOT = Path(__file__).resolve().parents[2]
DIMS = 384
SPARSE_QUERIES = 200
DENSE_QUERIES = 20
TOPK = 20


def sha(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()[:16]


def stride(n: int, want: int) -> list[int]:
    """等間隔。TS 側と同じ問いを見るため乱数を使わない。"""
    step = max(1, n // want)
    out = []
    i = 0
    while i < n and len(out) < want:
        out.append(i)
        i += step
    return out


def main() -> None:
    chunks = json.loads((ROOT / "data/chunks.json").read_text(encoding="utf-8"))["chunks"]
    items = json.loads((ROOT / "data/oracle.json").read_text(encoding="utf-8"))["items"]
    vecs = T.read_f32(str(ROOT / "data/index/vec_f32.bin"))
    qvecs = T.read_f32(str(ROOT / "data/index/query_f32.bin"))
    ids = [c["id"] for c in chunks]

    # --- 層 1: 分かち書き(網羅) ---
    token_hashes = {
        "chunks": [sha("".join(T.bigrams(c["indexText"]))) for c in chunks],
        "queries": [sha("".join(T.bigrams(it["query"]))) for it in items],
        "lengths": sha(",".join(str(T.token_length(c["indexText"])) for c in chunks)),
    }
    print(f"層1 分かち書き: チャンク {len(token_hashes['chunks'])} / クエリ {len(token_hashes['queries'])}")

    # --- 層 2: 疎索引 ---
    ix = T.BM25Index([(c["id"], c["indexText"]) for c in chunks])
    terms = sorted(ix.postings.keys())
    sparse_index = {
        "N": ix.N,
        "avgdl": ix.avgdl,
        "termCount": len(terms),
        "dlSum": sum(ix.dl),
        "termsHash": sha("".join(terms)),
        "dfHash": sha(",".join(str(len(ix.postings[t]) // 2) for t in terms)),
    }
    print(f"層2 疎索引: 語 {len(terms)} / 平均長 {ix.avgdl:.6f}")

    # --- 層 3: 検索順位 ---
    s_idx = stride(len(items), SPARSE_QUERIES)
    d_idx = stride(len(items), DENSE_QUERIES)
    sparse_ranks = []
    for n, i in enumerate(s_idx):
        it = items[i]
        hits = T.bm25_search(ix, it["query"], top_k=TOPK, exclude=set(it["excludeFromResults"]))
        sparse_ranks.append({"item": it["id"], "ids": [h["id"] for h in hits]})
        if n % 50 == 0:
            print(f"  疎 {n}/{len(s_idx)}")

    dense_ranks = []
    hybrid_ranks = []
    for n, i in enumerate(d_idx):
        it = items[i]
        exclude = set(it["excludeFromResults"])
        q = qvecs[i * DIMS : (i + 1) * DIMS]
        d = T.dense_search(vecs, DIMS, ids, q, top_k=TOPK, exclude=exclude)
        s = T.bm25_search(ix, it["query"], top_k=TOPK, exclude=exclude)
        dense_ranks.append(
            {"item": it["id"], "ids": [h["id"] for h in d], "topScore": d[0]["score"] if d else None}
        )
        hybrid_ranks.append({"item": it["id"], "ids": [h["id"] for h in T.rrf_fuse([d, s], top_k=TOPK)]})
        print(f"  密 {n + 1}/{len(d_idx)}")

    out = {
        "impl": "py",
        "tokenHashes": token_hashes,
        "sparseIndex": sparse_index,
        "sparseRanks": sparse_ranks,
        "denseRanks": dense_ranks,
        "hybridRanks": hybrid_ranks,
    }
    (ROOT / "data/ranks-py.json").write_text(json.dumps(out), encoding="utf-8")
    print("Python 側を書き出した → data/ranks-py.json")


if __name__ == "__main__":
    main()
